// cli-engine/lib/commands/commands.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _cliEngineCommand = require('cli-engine-command');

var _cliEngineCommand2 = _interopRequireDefault(_cliEngineCommand);

var _plugins = require('../plugins');

var _plugins2 = _interopRequireDefault(_plugins);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Commands extends _cliEngineCommand2.default {

  async run() {
    this.out.warn('heroku-cli: This CLI is deprecated. Please reinstall from https://cli.heroku.com');
    let plugins = new _plugins2.default(this.out);
    let topics = plugins.topics.filter(t => !t.hidden);
    let commands = plugins.commands.map(c => ({
      command: c.command,
      topic: c.topic,
      usage: c.usage,
      summary: c.description,
      help: c.help,
      hidden: c.hidden
    }));
    this.out.styledJSON({ topics, commands });
  }
}
exports.default = Commands;
Commands.topic = 'commands';
Commands.hidden = true;
Commands.flags = { json: _cliEngineCommand.flags.boolean() };
